[inductor] add the ability to do heavier search for coordinate descent tuning

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99403

When checking Meta's internal cmf10x model, I found this interesting kernel https://gist.github.com/shunting314/d4b1fc7352c840ef185c607392e21f31 . Doing coordinate descent tuning starting from the out of box tuner find sub-optimal config: a config worse than the best one max-autotuner can find.

This indicates that the coordinate descent tuner does not necessarily find the optimal config. Starting point matters.

I want to make the coordinate descent tuning less depend on the starting point. Also I think by improving that, the coordinate descent tuner may be more likely to find even better configs when starting from max-autotune result.

There are 2 ideas. 
1. currently coordinate descent tuning only considers changing one field/coordinate at a time. I add the ability to check all directions (i.e. tuning all tunable fields at the same time) after the normal coordinate descent searching does not find better choices. I'll check how that works in cmf10x
2. currently when we change a field, we only change 1 step (i.e. radius is 1). I add the ability to use a larger radius. This only affect the search in all directions and does not affect the normal coordinate descent searching workflow. 

Both are disabled by default.

Here are the tests I've done:


- OOB (out of the box): 0.083ms    0.003GB    38.13GB/s
- MA (max autotune): 0.016ms    0.003GB    195.60GB/s
   - best config: XBLOCK: 4, RBLOCK: 128, num_warps: 4, num_stages: 1

Default coordinate descent:
- Coordesc (coordinate descent tuner) upon OOB:  0.024ms    0.003GB    131.52GB/s ( **WORSE than Max Autotune** )
   - best config: XBLOCK: 64, RBLOCK: 4, num_warps: 16, num_stages: 1
- Coordesc upon MA: 0.016ms    0.003GB    194.31GB/s (no further improvement upon MA)

Search in all directions: (radius = 1)
- Coordesc upon OOB: 0.017ms    0.003GB    184.55GB/s 
   - best config: XBLOCK: 32, RBLOCK: 16, num_warps: 32, num_stages: 1
   - **IMPROVE FROM 0.024ms to 0.017ms. QUITE CLOSE TO THE ONE FIND BY MAX-AUTOTUNE**
- Coordesc upon MA: no further improvements upon MA

Search in all directions: (radius = 2)
- Coordesc upon OOB:  0.016ms    0.003GB    192.60GB/s
   - best config: XBLOCK: 8, RBLOCK: 16, num_warps: 8, num_stages: 1
   - **SLIGHTLY BETTER THAN RADIUS=1 for this kernel and on par with max-autotune**
- Coordesc upon MA: no further improvements upon MA


**Overall max-autotuner does a really good job for this kernel**





cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @aakhundov

shunting314 committed:
[inductor] add the ability to do heavier search for coordinate descent tuning

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] add the ability to do heavier search for coordinate descent tuning"


When checking Meta's internal cmf10x model, I found this interesting kernel https://gist.github.com/shunting314/d4b1fc7352c840ef185c607392e21f31 . Doing coordinate descent tuning starting from the out of box tuner find sub-optimal config: a config worse than the best one max-autotuner can find.

This indicates that the coordinate descent tuner does not necessarily find the optimal config. Starting point matters.

I want to make the coordinate descent tuning less depend on the starting point. Also I think by improving that, the coordinate descent tuner may be more likely to find even better configs when starting from max-autotune result.

There are 2 ideas. 
1. currently coordinate descent tuning only considers changing one field/coordinate at a time. I add the ability to check all directions (i.e. tuning all tunable fields at the same time) after the normal coordinate descent searching does not find better choices. I'll check how that works in cmf10x
2. currently when we change a field, we only change 1 step (i.e. radius is 1). I add the ability to use a larger radius. This only affect the search in all directions and does not affect the normal coordinate descent searching workflow. 

Both are disabled by default.

Here are the tests I've done:


- OOB (out of the box): 0.083ms    0.003GB    38.13GB/s
- MA (max autotune): 0.016ms    0.003GB    195.60GB/s
   - best config: XBLOCK: 4, RBLOCK: 128, num_warps: 4, num_stages: 1

Default coordinate descent:
- Coordesc (coordinate descent tuner) upon OOB:  0.024ms    0.003GB    131.52GB/s ( **WORSE than Max Autotune** )
   - best config: XBLOCK: 64, RBLOCK: 4, num_warps: 16, num_stages: 1
- Coordesc upon MA: 0.016ms    0.003GB    194.31GB/s (no further improvement upon MA)

Search in all directions: (radius = 1)
- Coordesc upon OOB: 0.017ms    0.003GB    184.55GB/s 
   - best config: XBLOCK: 32, RBLOCK: 16, num_warps: 32, num_stages: 1
   - **IMPROVE FROM 0.024ms to 0.017ms. QUITE CLOSE TO THE ONE FIND BY MAX-AUTOTUNE**
- Coordesc upon MA: no further improvements upon MA

Search in all directions: (radius = 2)
- Coordesc upon OOB:  0.016ms    0.003GB    192.60GB/s
   - best config: XBLOCK: 8, RBLOCK: 16, num_warps: 8, num_stages: 1
   - **SLIGHTLY BETTER THAN RADIUS=1 for this kernel and on par with max-autotune**
- Coordesc upon MA: no further improvements upon MA


**Overall max-autotuner does a really good job for this kernel**





[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] add the ability to do heavier search for coordinate descent tuning"


When checking Meta's internal cmf10x model, I found this interesting kernel https://gist.github.com/shunting314/d4b1fc7352c840ef185c607392e21f31 . Doing coordinate descent tuning starting from the out of box tuner find sub-optimal config: a config worse than the best one max-autotuner can find.

This indicates that the coordinate descent tuner does not necessarily find the optimal config. Starting point matters.

I want to make the coordinate descent tuning less depend on the starting point. Also I think by improving that, the coordinate descent tuner may be more likely to find even better configs when starting from max-autotune result.

There are 2 ideas. 
1. currently coordinate descent tuning only considers changing one field/coordinate at a time. I add the ability to check all directions (i.e. tuning all tunable fields at the same time) after the normal coordinate descent searching does not find better choices. I'll check how that works in cmf10x
2. currently when we change a field, we only change 1 step (i.e. radius is 1). I add the ability to use a larger radius. This only affect the search in all directions and does not affect the normal coordinate descent searching workflow. 

Both are disabled by default.

Here are the tests I've done:


- OOB (out of the box): 0.083ms    0.003GB    38.13GB/s
- MA (max autotune): 0.016ms    0.003GB    195.60GB/s
   - best config: XBLOCK: 4, RBLOCK: 128, num_warps: 4, num_stages: 1

Default coordinate descent:
- Coordesc (coordinate descent tuner) upon OOB:  0.024ms    0.003GB    131.52GB/s ( **WORSE than Max Autotune** )
   - best config: XBLOCK: 64, RBLOCK: 4, num_warps: 16, num_stages: 1
- Coordesc upon MA: 0.016ms    0.003GB    194.31GB/s (no further improvement upon MA)

Search in all directions: (radius = 1)
- Coordesc upon OOB: 0.017ms    0.003GB    184.55GB/s 
   - best config: XBLOCK: 32, RBLOCK: 16, num_warps: 32, num_stages: 1
   - **IMPROVE FROM 0.024ms to 0.017ms. QUITE CLOSE TO THE ONE FIND BY MAX-AUTOTUNE**
- Coordesc upon MA: no further improvements upon MA

Search in all directions: (radius = 2)
- Coordesc upon OOB:  0.016ms    0.003GB    192.60GB/s
   - best config: XBLOCK: 8, RBLOCK: 16, num_warps: 8, num_stages: 1
   - **SLIGHTLY BETTER THAN RADIUS=1 for this kernel and on par with max-autotune**
- Coordesc upon MA: no further improvements upon MA


**Overall max-autotuner does a really good job for this kernel**





[ghstack-poisoned]

jansel commented on torch/_inductor/coordinate_descent_tuner.py:
```diff
@@ -134,24 +236,36 @@ def autotune(
+                improved = self.check_all_dir(func)
+
+                if improved:
+                    import colorama
```
We should handle the case where this isn't installed.

shunting314 replied:
I can add a few util in util.py like 
- make_red_text , make_yellow_text
```
def make_red_text(text):
  if HAS_COLORAMA:
    return colorama.Fore.RED + text + colorama.Fore.RESET
  else:
    return text
```
they use colorama if availabe, and fallback to black color text otherwise.

jansel commented on torch/_inductor/coordinate_descent_tuner.py:
```diff
@@ -134,24 +236,36 @@ def autotune(
+
+                    msg = (
+                        colorama.Fore.RED
+                        + "LOOKING IN ALL DIR FOUND BETTER CONFIGS, IMPROVE %.3fx"
```
```suggestion
                        + "Coordinate descend tuning found improvement of %.3fx"
```

jansel commented on torch/_inductor/coordinate_descent_tuner.py:
```diff
@@ -80,27 +87,122 @@ def tunable_fields(self):
         threshold = 0.001  # 0.1%
         return test is not None and test < baseline * (1 - threshold)
 
+    def check_all_dir(
```
Full words in function names.  We should put autotune or tune in name.

jansel commented on torch/_inductor/coordinate_descent_tuner.py:
```diff
@@ -80,27 +87,122 @@ def tunable_fields(self):
+        # go thru different tuning choices for field
+        old_val = get_field(self.cur_config, field)
+
+        out = self.check_all_dir(func, next_field_idx + 1)
```
This use of recursion is a bit hard to read.  It took me a second to realize it was iterating over `tunable_fields` in reverse order.  I'd suggest a simple loop.

shunting314 replied:
A simple loop may be hard to work here since the number of nested loops (i.e. len(tunable_fields) ) is not fixed.

Let me give an example, let's say the tunable_fields are XBLOCK, RBLOCK, num_warps (other kernels may not have RBLOCK, or they may have a YBLOCK), assuming the current best config we found is XBLOCK=64, RBLOCK=64, num_warps=16

We want to explore all neighbor settings. The neighbor values for XBLOCK = [32, 64, 128], YBLOCK=[32, 64, 128], num_warps=[8, 16, 32] (this assumes a radius = 1, a larger radius just checks more neighbor values for each field).
Then we will do a cross product of all these candidate neighbor values. 
-  for each (XBLOCK, YBLOCK, num_warps) in  [32, 64, 128] x [32, 64, 128] x [8, 16, 32]

we will generate a config and benchmark it.

Let me know if this part need more comments :)

shunting314 replied:
NVM, I think I can use `itertools.product` to replace the backtracking here.

shunting314 committed:
Update on "[inductor] add the ability to do heavier search for coordinate descent tuning"


When checking Meta's internal cmf10x model, I found this interesting kernel https://gist.github.com/shunting314/d4b1fc7352c840ef185c607392e21f31 . Doing coordinate descent tuning starting from the out of box tuner find sub-optimal config: a config worse than the best one max-autotuner can find.

This indicates that the coordinate descent tuner does not necessarily find the optimal config. Starting point matters.

I want to make the coordinate descent tuning less depend on the starting point. Also I think by improving that, the coordinate descent tuner may be more likely to find even better configs when starting from max-autotune result.

There are 2 ideas. 
1. currently coordinate descent tuning only considers changing one field/coordinate at a time. I add the ability to check all directions (i.e. tuning all tunable fields at the same time) after the normal coordinate descent searching does not find better choices. I'll check how that works in cmf10x
2. currently when we change a field, we only change 1 step (i.e. radius is 1). I add the ability to use a larger radius. This only affect the search in all directions and does not affect the normal coordinate descent searching workflow. 

Both are disabled by default.

Here are the tests I've done:


- OOB (out of the box): 0.083ms    0.003GB    38.13GB/s
- MA (max autotune): 0.016ms    0.003GB    195.60GB/s
   - best config: XBLOCK: 4, RBLOCK: 128, num_warps: 4, num_stages: 1

Default coordinate descent:
- Coordesc (coordinate descent tuner) upon OOB:  0.024ms    0.003GB    131.52GB/s ( **WORSE than Max Autotune** )
   - best config: XBLOCK: 64, RBLOCK: 4, num_warps: 16, num_stages: 1
- Coordesc upon MA: 0.016ms    0.003GB    194.31GB/s (no further improvement upon MA)

Search in all directions: (radius = 1)
- Coordesc upon OOB: 0.017ms    0.003GB    184.55GB/s 
   - best config: XBLOCK: 32, RBLOCK: 16, num_warps: 32, num_stages: 1
   - **IMPROVE FROM 0.024ms to 0.017ms. QUITE CLOSE TO THE ONE FIND BY MAX-AUTOTUNE**
- Coordesc upon MA: no further improvements upon MA

Search in all directions: (radius = 2)
- Coordesc upon OOB:  0.016ms    0.003GB    192.60GB/s
   - best config: XBLOCK: 8, RBLOCK: 16, num_warps: 8, num_stages: 1
   - **SLIGHTLY BETTER THAN RADIUS=1 for this kernel and on par with max-autotune**
- Coordesc upon MA: no further improvements upon MA


**Overall max-autotuner does a really good job for this kernel**





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] add the ability to do heavier search for coordinate descent tuning"


When checking Meta's internal cmf10x model, I found this interesting kernel https://gist.github.com/shunting314/d4b1fc7352c840ef185c607392e21f31 . Doing coordinate descent tuning starting from the out of box tuner find sub-optimal config: a config worse than the best one max-autotuner can find.

This indicates that the coordinate descent tuner does not necessarily find the optimal config. Starting point matters.

I want to make the coordinate descent tuning less depend on the starting point. Also I think by improving that, the coordinate descent tuner may be more likely to find even better configs when starting from max-autotune result.

There are 2 ideas. 
1. currently coordinate descent tuning only considers changing one field/coordinate at a time. I add the ability to check all directions (i.e. tuning all tunable fields at the same time) after the normal coordinate descent searching does not find better choices. I'll check how that works in cmf10x
2. currently when we change a field, we only change 1 step (i.e. radius is 1). I add the ability to use a larger radius. This only affect the search in all directions and does not affect the normal coordinate descent searching workflow. 

Both are disabled by default.

Here are the tests I've done:


- OOB (out of the box): 0.083ms    0.003GB    38.13GB/s
- MA (max autotune): 0.016ms    0.003GB    195.60GB/s
   - best config: XBLOCK: 4, RBLOCK: 128, num_warps: 4, num_stages: 1

Default coordinate descent:
- Coordesc (coordinate descent tuner) upon OOB:  0.024ms    0.003GB    131.52GB/s ( **WORSE than Max Autotune** )
   - best config: XBLOCK: 64, RBLOCK: 4, num_warps: 16, num_stages: 1
- Coordesc upon MA: 0.016ms    0.003GB    194.31GB/s (no further improvement upon MA)

Search in all directions: (radius = 1)
- Coordesc upon OOB: 0.017ms    0.003GB    184.55GB/s 
   - best config: XBLOCK: 32, RBLOCK: 16, num_warps: 32, num_stages: 1
   - **IMPROVE FROM 0.024ms to 0.017ms. QUITE CLOSE TO THE ONE FIND BY MAX-AUTOTUNE**
- Coordesc upon MA: no further improvements upon MA

Search in all directions: (radius = 2)
- Coordesc upon OOB:  0.016ms    0.003GB    192.60GB/s
   - best config: XBLOCK: 8, RBLOCK: 16, num_warps: 8, num_stages: 1
   - **SLIGHTLY BETTER THAN RADIUS=1 for this kernel and on par with max-autotune**
- Coordesc upon MA: no further improvements upon MA


**Overall max-autotuner does a really good job for this kernel**





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

shunting314 commented:
Sorry for being late. Finally get a chance to followup on the review comments of this PR. @jansel  please take another look when you get a chance.

shunting314 committed:
Update on "[inductor] add the ability to do heavier search for coordinate descent tuning"


When checking Meta's internal cmf10x model, I found this interesting kernel https://gist.github.com/shunting314/d4b1fc7352c840ef185c607392e21f31 . Doing coordinate descent tuning starting from the out of box tuner find sub-optimal config: a config worse than the best one max-autotuner can find.

This indicates that the coordinate descent tuner does not necessarily find the optimal config. Starting point matters.

I want to make the coordinate descent tuning less depend on the starting point. Also I think by improving that, the coordinate descent tuner may be more likely to find even better configs when starting from max-autotune result.

There are 2 ideas. 
1. currently coordinate descent tuning only considers changing one field/coordinate at a time. I add the ability to check all directions (i.e. tuning all tunable fields at the same time) after the normal coordinate descent searching does not find better choices. I'll check how that works in cmf10x
2. currently when we change a field, we only change 1 step (i.e. radius is 1). I add the ability to use a larger radius. This only affect the search in all directions and does not affect the normal coordinate descent searching workflow. 

Both are disabled by default.

Here are the tests I've done:


- OOB (out of the box): 0.083ms    0.003GB    38.13GB/s
- MA (max autotune): 0.016ms    0.003GB    195.60GB/s
   - best config: XBLOCK: 4, RBLOCK: 128, num_warps: 4, num_stages: 1

Default coordinate descent:
- Coordesc (coordinate descent tuner) upon OOB:  0.024ms    0.003GB    131.52GB/s ( **WORSE than Max Autotune** )
   - best config: XBLOCK: 64, RBLOCK: 4, num_warps: 16, num_stages: 1
- Coordesc upon MA: 0.016ms    0.003GB    194.31GB/s (no further improvement upon MA)

Search in all directions: (radius = 1)
- Coordesc upon OOB: 0.017ms    0.003GB    184.55GB/s 
   - best config: XBLOCK: 32, RBLOCK: 16, num_warps: 32, num_stages: 1
   - **IMPROVE FROM 0.024ms to 0.017ms. QUITE CLOSE TO THE ONE FIND BY MAX-AUTOTUNE**
- Coordesc upon MA: no further improvements upon MA

Search in all directions: (radius = 2)
- Coordesc upon OOB:  0.016ms    0.003GB    192.60GB/s
   - best config: XBLOCK: 8, RBLOCK: 16, num_warps: 8, num_stages: 1
   - **SLIGHTLY BETTER THAN RADIUS=1 for this kernel and on par with max-autotune**
- Coordesc upon MA: no further improvements upon MA


**Overall max-autotuner does a really good job for this kernel**





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge
